## 1. Data Understanding

In [1]:
import pandas as pd
import numpy as np
df_2009_2010 = pd.read_excel('../data/raw/online_retail_II.xlsx', sheet_name='Year 2009-2010')
df_2010_2011 = pd.read_excel('../data/raw/online_retail_II.xlsx', sheet_name='Year 2010-2011')
df = pd.concat([df_2009_2010, df_2010_2011], ignore_index=True)

df.to_csv('../data/raw/online_retail_combined.csv', index=False)

In [2]:
print(df.shape)
df.head()

(1067371, 8)


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [3]:
df. info

<bound method DataFrame.info of         Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067370  581587      POST                              POSTAGE         1   

                InvoiceDate  Price  Customer ID        

In [4]:
df. duplicated().sum()

np.int64(34335)

In [5]:
df .isnull(). sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [6]:
missing = df .isnull(). sum()
missing_pct = (missing / len(df)) *100
pd.DataFrame({'missing_count': missing, 'missing_pct': missing_pct.round(2)})

,missing_count,missing_pct
Invoice,0,0.00
StockCode,0,0.00
Description,4382,0.41
Quantity,0,0.00
InvoiceDate,0,0.00
Price,0,0.00
Customer ID,243007,22.77
Country,0,0.00


In [7]:
print(df.columns.tolist())

['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']


In [8]:
df[['Quantity', 'Price']].describe()

,Quantity,Price
count,1.067371e+06,1.067371e+06
mean,9.938898e+00,4.649388e+00
std,1.727058e+02,1.235531e+02
min,-8.099500e+04,-5.359436e+04
25%,1.000000e+00,1.250000e+00
50%,3.000000e+00,2.100000e+00
75%,1.000000e+01,4.150000e+00
max,8.099500e+04,3.897000e+04


In [9]:
df['InvoiceDate'].min() , df['InvoiceDate'].max()

(Timestamp('2009-12-01 07:45:00'), Timestamp('2011-12-09 12:50:00'))

In [10]:
for col in ['Customer ID', 'StockCode', 'Invoice', 'Country']:
    print(col, '->', df[col].nunique())

Customer ID -> 5942
StockCode -> 5305
Invoice -> 53628
Country -> 43


In [11]:
cancelled = df[df['Invoice'].astype(str).str.startswith('C')]
print(f"{len(cancelled)} cancelled transactions, {len(cancelled)/len(df)*100:.2f}% of all rows")

19494 cancelled transactions, 1.83% of all rows


In [12]:
(df['Price'] <= 0).sum()

np.int64(6207)

In [13]:
df[['Quantity', 'Price']].describe()

,Quantity,Price
count,1.067371e+06,1.067371e+06
mean,9.938898e+00,4.649388e+00
std,1.727058e+02,1.235531e+02
min,-8.099500e+04,-5.359436e+04
25%,1.000000e+00,1.250000e+00
50%,3.000000e+00,2.100000e+00
75%,1.000000e+01,4.150000e+00
max,8.099500e+04,3.897000e+04


In [14]:
df[['Description', 'Country', 'Invoice', 'StockCode']].describe()

,Description,Country,Invoice,StockCode
count,1062989,1067371,1067371,1067371
unique,5698,43,53628,5305
top,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom,537434,85123A
freq,5918,981330,1350,5829


In [15]:
df['Country'].value_counts()

Country
United Kingdom          981330
EIRE                     17866
Germany                  17624
France                   14330
Netherlands               5140
Spain                     3811
Switzerland               3189
Belgium                   3123
Portugal                  2620
Australia                 1913
Channel Islands           1664
Italy                     1534
Norway                    1455
Sweden                    1364
Cyprus                    1176
Finland                   1049
Austria                    938
Denmark                    817
Unspecified                756
Greece                     663
Japan                      582
Poland                     535
USA                        535
United Arab Emirates       500
Israel                     371
Hong Kong                  364
Singapore                  346
Malta                      299
Iceland                    253
Canada                     228
Lithuania                  189
RSA                        169


In [16]:
# Top 20 most frequent products
df['Description'].value_counts().head(20)

Description
WHITE HANGING HEART T-LIGHT HOLDER    5918
REGENCY CAKESTAND 3 TIER              4412
JUMBO BAG RED RETROSPOT               3469
ASSORTED COLOUR BIRD ORNAMENT         2958
PARTY BUNTING                         2765
STRAWBERRY CERAMIC TRINKET BOX        2613
LUNCH BAG  BLACK SKULL.               2529
JUMBO STORAGE BAG SUKI                2434
HEART OF WICKER SMALL                 2319
JUMBO SHOPPER VINTAGE RED PAISLEY     2297
60 TEATIME FAIRY CAKE CASES           2271
PAPER CHAIN KIT 50'S CHRISTMAS        2215
LUNCH BAG SPACEBOY DESIGN             2206
REX CASH+CARRY JUMBO SHOPPER          2199
HOME BUILDING BLOCK WORD              2192
WOODEN FRAME ANTIQUE WHITE            2189
LUNCH BAG CARS BLUE                   2184
NATURAL SLATE HEART CHALKBOARD        2151
BAKING SET 9 PIECE RETROSPOT          2150
POSTAGE                               2115
Name: count, dtype: int64

In [17]:
df['Country'].value_counts(normalize=True) * 100

Country
United Kingdom          91.938979
EIRE                     1.673832
Germany                  1.651160
France                   1.342551
Netherlands              0.481557
Spain                    0.357045
Switzerland              0.298771
Belgium                  0.292588
Portugal                 0.245463
Australia                0.179225
Channel Islands          0.155897
Italy                    0.143718
Norway                   0.136316
Sweden                   0.127791
Cyprus                   0.110177
Finland                  0.098279
Austria                  0.087879
Denmark                  0.076543
Unspecified              0.070828
Greece                   0.062115
Japan                    0.054526
Poland                   0.050123
USA                      0.050123
United Arab Emirates     0.046844
Israel                   0.034758
Hong Kong                0.034102
Singapore                0.032416
Malta                    0.028013
Iceland                  0.023703
Canada

In [18]:
df['StockCode'].value_counts().head(20)

StockCode
85123A    5829
22423     4424
85099B    4216
21212     3318
20725     3259
84879     2960
47566     2768
21232     2747
22197     2549
22383     2540
20727     2529
21931     2434
22386     2347
22469     2325
22411     2297
84991     2271
22382     2251
22384     2230
21080     2224
22086     2217
Name: count, dtype: int64

In [19]:
df['Invoice'].value_counts().describe()

count    53628.000000
mean        19.903241
std         41.252573
min          1.000000
25%          2.000000
50%          9.000000
75%         23.000000
max       1350.000000
Name: count, dtype: float64

In [20]:
# Negative Quantity (returns/cancellations)
neg_qty = (df['Quantity'] < 0).sum()
zero_qty = (df['Quantity'] == 0).sum()
print(f"Negative Quantity: {neg_qty} rows ({neg_qty/len(df)*100:.2f}%)")
print(f"Zero Quantity: {zero_qty} rows ({zero_qty/len(df)*100:.2f}%)")

Negative Quantity: 22950 rows (2.15%)
Zero Quantity: 0 rows (0.00%)


In [21]:
# Negative and zero Price
neg_price = (df['Price'] < 0).sum()
zero_price = (df['Price'] == 0).sum()
print(f"Negative Price: {neg_price} rows ({neg_price/len(df)*100:.2f}%)")
print(f"Zero Price: {zero_price} rows ({zero_price/len(df)*100:.2f}%)")

Negative Price: 5 rows (0.00%)
Zero Price: 6202 rows (0.58%)


In [22]:
# Look at the actual rows with the most extreme Quantity values
df.nlargest(10, 'Quantity')[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'Customer ID']]
df.nsmallest(10, 'Quantity')[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'Customer ID']]

# Same for Price
df.nlargest(10, 'Price')[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'Customer ID']]

,Invoice,StockCode,Description,Quantity,Price,Customer ID
748142,C556445,M,Manual,-1,38970.00,15098.0
241824,C512770,M,Manual,-1,25111.09,17399.0
241827,512771,M,Manual,1,25111.09,NaN
320581,C520667,BANK CHARGES,Bank Charges,-1,18910.69,NaN
1050063,C580605,AMAZONFEE,AMAZON FEE,-1,17836.46,NaN
569163,C540117,AMAZONFEE,AMAZON FEE,-1,16888.02,NaN
569164,C540118,AMAZONFEE,AMAZON FEE,-1,16453.71,NaN
517953,C537630,AMAZONFEE,AMAZON FEE,-1,13541.33,NaN
517955,537632,AMAZONFEE,AMAZON FEE,1,13541.33,NaN
519294,C537651,AMAZONFEE,AMAZON FEE,-1,13541.33,NaN


In [23]:
# Find all non-numeric-looking StockCodes — real products are typically 5-digit codes, sometimes with a letter suffix
for code in ['M', 'BANK CHARGES', 'AMAZONFEE', 'POST', 'DOT', 'D', 'CRUK', 'C2', 'CRUK', 'B', 'ADJUST', 'ADJUST2']:
    count = (df['StockCode'] == code).sum()
    if count > 0:
        print(f"{code}: {count} rows")       

M: 1421 rows
BANK CHARGES: 102 rows
AMAZONFEE: 43 rows
POST: 2122 rows
DOT: 1446 rows
D: 177 rows
CRUK: 16 rows
C2: 282 rows
CRUK: 16 rows
B: 6 rows
ADJUST: 67 rows
ADJUST2: 3 rows


In [24]:
df.nlargest(5, 'Quantity')[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'Customer ID']]

,Invoice,StockCode,Description,Quantity,Price,Customer ID
1065882,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2.08,16446.0
587080,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,1.04,12346.0
90857,497946,37410,BLACK AND WHITE PAISLEY FLOWER MUG,19152,0.10,13902.0
127166,501534,21099,SET/6 STRAWBERRY PAPER CUPS,12960,0.10,13902.0
127168,501534,21091,SET/6 WOODLAND PAPER PLATES,12960,0.10,13902.0


In [25]:
df.nsmallest(5, 'Quantity')[['Invoice', 'StockCode', 'Description', 'Quantity', 'Price', 'Customer ID']]

,Invoice,StockCode,Description,Quantity,Price,Customer ID
1065883,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2.08,16446.0
587085,C541433,23166,MEDIUM CERAMIC TOP STORAGE JAR,-74215,1.04,12346.0
303996,519017,22759,NaN,-9600,0.00,NaN
750990,556690,23005,printing smudges/thrown away,-9600,0.00,NaN
750991,556691,23005,printing smudges/thrown away,-9600,0.00,NaN


The combined dataset contains 1,067,371 rows across 8 columns, spanning Dec 2009–Dec 2011. The business is heavily UK-concentrated: 91.9% of transactions originate from the United Kingdom, with the remaining 8.1% spread across 42 other countries, plus 756 rows (0.07%) labeled "Unspecified." Customer ID is missing in 22.7% of rows — these transactions still represent real revenue but can't be attributed to a specific customer. 2.15% of rows have negative Quantity (22,950 rows), corresponding to cancellations/returns; Quantity also contains extreme outliers (min -80,995, max 80,995), which will require investigation beyond a simple sign check. Price has 5 negative values (negligible) and 6,202 zero values (0.58%), likely representing free items or adjustments; Price also has a right-skewed outlier at its max (38,970) relative to its 75th percentile (4.15). The "POSTAGE" line item appears 2,115 times as a top "product," representing shipping charges rather than real merchandise and requiring separate handling. Median items per invoice is 9, but with a long tail up to 1,350 items on a single invoice, consistent with a mix of individual retail purchases and bulk wholesale orders.

### 2. Data Cleaning

In [26]:
#Convert data types
df ['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])
df ['StockCode'] = df['StockCode'].astype(str)
df ['Invoice'] = df['Invoice'].astype(str)

In [27]:
#Handle missing rows
missing_customer = df['Customer ID'].isnull().sum()
print(f"{missing_customer} rows ({missing_customer/len(df)*100: .2f}%) missing Customer ID")
df_with_customer = df.dropna(subset=['Customer ID']).copy()
df_with_customer['Customer ID'] = df_with_customer['Customer ID'].astype(int)                                                         

243007 rows ( 22.77%) missing Customer ID


In [28]:
#Seperate cancellations 
df['is_cancelled'] = df['Invoice'].str.startswith('C')
cancelled_count = df['is_cancelled'].sum()
df_sales = df[~df['is_cancelled']].copy()
df_cancellations = df[df['is_cancelled']].copy()

In [29]:
#Remove non product administrative codes 
non_product_codes = ['POST', 'DOT', 'M', 'C2', 'D', 'BANK CHARGES', 'AMAZONFEE', 'CRUK', 'B', 'ADJUST', 'ADJUST2']

removed_admin = df_sales['StockCode'].isin(non_product_codes).sum()
print(f"{removed_admin} rows removed as non-product administrative entries")

df_sales = df_sales[~df_sales['StockCode'].isin(non_product_codes)].copy()

4583 rows removed as non-product administrative entries


In [30]:
#Remove zero/negative Price
invalid_price = (df_sales['Price'] <= 0).sum()
print(f"{invalid_price} rows with zero or negative Price removed")

df_sales = df_sales[df_sales['Price'] > 0].copy()

6178 rows with zero or negative Price removed


In [31]:
#Handle missing Description
missing_desc = df_sales['Description'].isnull().sum()
print(f"{missing_desc} rows with missing Description filled as 'Unknow Product'")

df_sales['Description'] = df_sales['Description'].fillna('Unknown Product')

0 rows with missing Description filled as 'Unknow Product'


In [32]:
#remove duplicate rows
dupes = df_sales.duplicated().sum()
print(f"{dupes} exact duplicate rows removed")

df_sales = df_sales.drop_duplicates().copy()

33664 exact duplicate rows removed


In [33]:
#confirm no negative Quantity slipped through
remaining_negative_qty = (df_sales ['Quantity'] < 0).sum()
print(f"Remaining negative Quantity rows in df_sales: {remaining_negative_qty}")

Remaining negative Quantity rows in df_sales: 0


In [34]:
#Create Revenue column
df_sales['Revenue'] = df_sales['Quantity'] * df_sales['Price']

## 3. Feature Engineering: Product Category

In [35]:
df_sales['Description'].value_counts().head(60)

Description
WHITE HANGING HEART T-LIGHT HOLDER     5601
REGENCY CAKESTAND 3 TIER               3950
JUMBO BAG RED RETROSPOT                3312
ASSORTED COLOUR BIRD ORNAMENT          2848
PARTY BUNTING                          2705
LUNCH BAG  BLACK SKULL.                2402
STRAWBERRY CERAMIC TRINKET BOX         2352
JUMBO STORAGE BAG SUKI                 2352
JUMBO SHOPPER VINTAGE RED PAISLEY      2209
HEART OF WICKER SMALL                  2195
60 TEATIME FAIRY CAKE CASES            2179
LUNCH BAG SPACEBOY DESIGN              2128
LUNCH BAG CARS BLUE                    2091
WOODEN FRAME ANTIQUE WHITE             2087
HOME BUILDING BLOCK WORD               2087
PAPER CHAIN KIT 50'S CHRISTMAS         2068
NATURAL SLATE HEART CHALKBOARD         2059
WOODEN PICTURE FRAME WHITE FINISH      2026
PACK OF 60 PINK PAISLEY CAKE CASES     2025
REX CASH+CARRY JUMBO SHOPPER           1996
HEART OF WICKER LARGE                  1994
LUNCH BAG WOODLAND                     1950
BAKING SET 9 PIECE R

In [36]:
def categorize_product(description):
    desc = str(description).upper()

    if any(word in desc for word in ['CHRISTMAS', 'XMAS', 'ADVENT']):
        return 'Christmas & Seasonal'
    elif any(word in desc for word in ['BAG', 'SHOPPER', 'BASKET']):
      return 'Bags & Storage'
    elif any(word in desc for word in ['CAKE', 'BAKING', 'TEATIME', 'MUG', 'TEACUP', 'SAUCER', 'SNACK BOX', 'HOT WATER BOTTLE']):
      return 'Kitchen & Dining '
    elif any(word in desc for word in ['LIGHT HOLDER', 'T-LIGHT', 'CANDLE', 'ORNAMENT', 'FRAME', 'CHALKBOARD', 'NIGHT LIGHT', 'DECORATION']):
      return 'Home Decor & Lighting'
    elif any(word in desc for word in ['SIGNS']):
      return 'Novelty Signs'
    elif any(word in desc for word in ['BUNTING', 'PARTY', 'BALLOON']):
      return 'Party & Occasion'
    elif any(word in desc for word in ['TRINKET BOX', 'RECIPE BOX', 'BOX']):
        return 'Storage & Boxes'
    elif any(word in desc for word in ['CARD', 'WRAP', 'GIFT TAG', 'RIBBON', 'PAPER CHAIN'] ):
        return 'Stationery & Gift Wrap'
    elif any(word in desc for word in['FELT', 'CRAFT', 'WICKER']):
        return 'Craft & Textile'
    else:
        return 'Other'      

df_sales['Category'] = df_sales['Description'].apply(categorize_product)

In [37]:
df_sales['Category'].value_counts()

Category
Other                     512100
Kitchen & Dining          100326
Home Decor & Lighting     100018
Bags & Storage             99618
Stationery & Gift Wrap     51880
Storage & Boxes            46573
Christmas & Seasonal       43692
Craft & Textile            26058
Party & Occasion           22165
Novelty Signs               1022
Name: count, dtype: int64

In [38]:
df_sales['Category'].value_counts(normalize=True) * 100

Category
Other                     51.033831
Kitchen & Dining           9.998087
Home Decor & Lighting      9.967393
Bags & Storage             9.927530
Stationery & Gift Wrap     5.170153
Storage & Boxes            4.641278
Christmas & Seasonal       4.354169
Craft & Textile            2.596836
Party & Occasion           2.208875
Novelty Signs              0.101848
Name: proportion, dtype: float64

In [39]:
df_sales[df_sales['Category'] == 'Other']['Description'].value_counts().head(40)

Description
HOME BUILDING BLOCK WORD               2087
LOVE BUILDING BLOCK WORD               1717
PLEASE ONE PERSON METAL SIGN           1649
GIN + TONIC DIET METAL SIGN            1621
HAND OVER THE CHOCOLATE   SIGN         1472
COOK WITH WINE METAL SIGN              1435
SMALL POPCORN HOLDER                   1428
ALARM CLOCK BAKELIKE RED               1380
BATHROOM METAL SIGN                    1349
RETROSPOT TEA SET CERAMIC 11 PC        1326
PLASTERS IN TIN SPACEBOY               1314
PLASTERS IN TIN WOODLAND ANIMALS       1292
ALARM CLOCK BAKELIKE GREEN             1285
SET OF 3 HEART COOKIE CUTTERS          1264
HAND WARMER OWL DESIGN                 1256
IVORY KITCHEN SCALES                   1221
PLASTERS IN TIN CIRCUS PARADE          1216
CLOTHES PEGS RETROSPOT PACK 24         1212
JAM MAKING SET PRINTED                 1203
SET/20 RED RETROSPOT PAPER NAPKINS     1202
HOT BATHS METAL SIGN                   1198
WOOD BLACK BOARD ANT WHITE FINISH      1177
SET/5 RED RETROSPOT 

In [40]:
df_sales['Category'] = df_sales['Description'].apply(categorize_product)
df_sales[df_sales['Category'] == 'Other']['Description'].value_counts().head(40)

Description
HOME BUILDING BLOCK WORD               2087
LOVE BUILDING BLOCK WORD               1717
PLEASE ONE PERSON METAL SIGN           1649
GIN + TONIC DIET METAL SIGN            1621
HAND OVER THE CHOCOLATE   SIGN         1472
COOK WITH WINE METAL SIGN              1435
SMALL POPCORN HOLDER                   1428
ALARM CLOCK BAKELIKE RED               1380
BATHROOM METAL SIGN                    1349
RETROSPOT TEA SET CERAMIC 11 PC        1326
PLASTERS IN TIN SPACEBOY               1314
PLASTERS IN TIN WOODLAND ANIMALS       1292
ALARM CLOCK BAKELIKE GREEN             1285
SET OF 3 HEART COOKIE CUTTERS          1264
HAND WARMER OWL DESIGN                 1256
IVORY KITCHEN SCALES                   1221
PLASTERS IN TIN CIRCUS PARADE          1216
CLOTHES PEGS RETROSPOT PACK 24         1212
JAM MAKING SET PRINTED                 1203
SET/20 RED RETROSPOT PAPER NAPKINS     1202
HOT BATHS METAL SIGN                   1198
WOOD BLACK BOARD ANT WHITE FINISH      1177
SET/5 RED RETROSPOT 

In [41]:
def categorize_product(description):
    desc = str(description).upper()
    
    if any(word in desc for word in ['CHRISTMAS', 'XMAS', 'ADVENT']):
        return 'Christmas & Seasonal'
    elif any(word in desc for word in ['BAG', 'SHOPPER', 'BASKET']):
        return 'Bags & Storage'
    elif any(word in desc for word in ['CAKE', 'BAKING', 'TEATIME', 'MUG', 'TEACUP', 'SAUCER',
                                         'SNACK BOX', 'HOT WATER BOTTLE', 'KITCHEN SCALES',
                                         'JAM MAKING', 'COOKIE CUTTER', 'NAPKIN', 'LID GLASS BOWL',
                                         'JELLY MOULD', 'TEA SET', 'TEA GLASS']):
        return 'Kitchen & Dining'
    elif any(word in desc for word in ['LIGHT HOLDER', 'T-LIGHT', 'CANDLE', 'ORNAMENT', 'FRAME',
                                         'CHALKBOARD', 'NIGHT LIGHT', 'DECORATION', 'CHILLI LIGHTS']):
        return 'Home Decor & Lighting'
    elif 'SIGN' in desc:
        return 'Novelty Signs'
    elif any(word in desc for word in ['BUNTING', 'PARTY', 'BALLOON']):
        return 'Party & Occasion'
    elif any(word in desc for word in ['TRINKET BOX', 'RECIPE BOX', 'TIN', 'BOX', 'CABINET', 'COAT RACK']):
        return 'Storage & Boxes'
    elif any(word in desc for word in ['CARD', 'WRAP', 'GIFT TAG', 'RIBBON', 'PAPER CHAIN']):
        return 'Stationery & Gift Wrap'
    elif any(word in desc for word in ['FELT', 'CRAFT', 'WICKER']):
        return 'Craft & Textile'
    elif any(word in desc for word in ['BUILDING BLOCK', 'ALARM CLOCK', 'HAND WARMER', 'PLASTERS',
                                         'PARASOL', 'PEG']):
        return 'Novelty & Personal Items'
    else:
        return 'Other'

df_sales['Category'] = df_sales['Description'].apply(categorize_product)
df_sales['Category'].value_counts(normalize=True) * 100

Category
Other                       38.171532
Kitchen & Dining            12.573098
Home Decor & Lighting       10.081100
Bags & Storage               9.927530
Storage & Boxes              7.587508
Novelty Signs                6.117084
Stationery & Gift Wrap       4.535743
Christmas & Seasonal         4.354169
Craft & Textile              2.596836
Party & Occasion             2.103838
Novelty & Personal Items     1.951563
Name: proportion, dtype: float64

In [42]:
df_sales[df_sales['Category'] == 'Other']['Description'].value_counts().head(40)

Description
SMALL POPCORN HOLDER                  1428
WOOD BLACK BOARD ANT WHITE FINISH     1177
RED WOOLLY HOTTIE WHITE HEART.        1163
SMALL GLASS HEART TRINKET POT         1072
DOORMAT HEARTS                        1035
SET OF 72 PINK HEART PAPER DOILIES    1023
METAL 4 HOOK HANGER FRENCH CHATEAU    1006
SMALL HEART MEASURING SPOONS          1005
DOORMAT UNION FLAG                     994
ASSORTED COLOURS SILK FAN              992
SET/6 RED SPOTTY PAPER PLATES          990
HEART IVORY TRELLIS SMALL              989
CREAM SWEETHEART MINI CHEST            938
WOODEN SCHOOL COLOURING SET            922
GARDENERS KNEELING PAD KEEP CALM       920
DOORMAT RED RETROSPOT                  901
WHITE WOOD GARDEN PLANT LADDER         900
HEART FILIGREE DOVE  SMALL             894
TRADITIONAL WOODEN SKIPPING ROPE       859
4 TRADITIONAL SPINNING TOPS            857
SET/6 RED SPOTTY PAPER CUPS            854
HEART IVORY TRELLIS LARGE              854
CREAM CUPID HEARTS COAT HANGER         853

In [43]:
def categorize_product(description):
    desc = str(description).upper()
    
    if any(word in desc for word in ['CHRISTMAS', 'XMAS', 'ADVENT']):
        return 'Christmas & Seasonal'
    elif any(word in desc for word in ['BAG', 'SHOPPER', 'BASKET']):
        return 'Bags & Storage'
    elif any(word in desc for word in ['CAKE', 'BAKING', 'TEATIME', 'MUG', 'TEACUP', 'SAUCER',
                                         'SNACK BOX', 'HOT WATER BOTTLE', 'KITCHEN SCALES',
                                         'JAM MAKING', 'COOKIE CUTTER', 'NAPKIN', 'LID GLASS BOWL',
                                         'JELLY MOULD', 'TEA SET', 'TEA GLASS', 'MEASURING SPOONS',
                                         'MILK JUG', 'DOILIES']):
        return 'Kitchen & Dining'
    elif any(word in desc for word in ['LIGHT HOLDER', 'T-LIGHT', 'CANDLE', 'ORNAMENT', 'FRAME',
                                         'CHALKBOARD', 'NIGHT LIGHT', 'DECORATION', 'CHILLI LIGHTS',
                                         'POPCORN HOLDER']):
        return 'Home Decor & Lighting'
    elif 'SIGN' in desc:
        return 'Novelty Signs'
    elif any(word in desc for word in ['BUNTING', 'PARTY', 'BALLOON']):
        return 'Party & Occasion'
    elif any(word in desc for word in ['TRINKET', 'RECIPE BOX', 'TIN', 'BOX', 'CABINET', 'COAT RACK']):
        return 'Storage & Boxes'
    elif any(word in desc for word in ['CARD', 'WRAP', 'GIFT TAG', 'RIBBON', 'PAPER CHAIN',
                                         'NOTEBOOK', 'TISSUE', 'PENCIL']):
        return 'Stationery & Gift Wrap'
    elif any(word in desc for word in ['FELT', 'CRAFT', 'WICKER']):
        return 'Craft & Textile'
    elif any(word in desc for word in ['BUILDING BLOCK', 'ALARM CLOCK', 'HAND WARMER', 'PLASTERS',
                                         'PARASOL', 'PEG', 'KEY FOB', 'SEWING KIT', 'FAN']):
        return 'Novelty & Personal Items'
    elif any(word in desc for word in ['DOORMAT', 'GARDEN', 'PLANT LADDER', 'KNEELING PAD']):
        return 'Garden & Outdoor'
    elif any(word in desc for word in ['HOOK', 'HANGER', 'DRAWER KNOB']):
        return 'Hardware & Fixtures'
    elif any(word in desc for word in ['SPINNING TOP', 'SKIPPING ROPE', 'SNAKES & LADDERS',
                                         'BINGO', 'CATCH CUP', 'TOY', 'GAME']):
        return 'Toys & Games'
    else:
        return 'Other'

df_sales['Category'] = df_sales['Description'].apply(categorize_product)
df_sales['Category'].value_counts(normalize=True) * 100

Category
Other                       28.618011
Kitchen & Dining            13.513053
Home Decor & Lighting       10.382061
Bags & Storage               9.927530
Storage & Boxes              7.884782
Stationery & Gift Wrap       6.764449
Novelty Signs                6.075926
Christmas & Seasonal         4.354169
Novelty & Personal Items     2.742034
Craft & Textile              2.596836
Garden & Outdoor             2.468479
Party & Occasion             2.103838
Hardware & Fixtures          1.704516
Toys & Games                 0.864316
Name: proportion, dtype: float64

In [44]:
df_sales[df_sales['Category'] == 'Other']['Description'].value_counts().head(40)

Description
WOOD BLACK BOARD ANT WHITE FINISH    1177
RED WOOLLY HOTTIE WHITE HEART.       1163
SET/6 RED SPOTTY PAPER PLATES         990
HEART IVORY TRELLIS SMALL             989
CREAM SWEETHEART MINI CHEST           938
WOODEN SCHOOL COLOURING SET           922
HEART FILIGREE DOVE  SMALL            894
SET/6 RED SPOTTY PAPER CUPS           854
HEART IVORY TRELLIS LARGE             854
ENAMEL FLOWER JUG CREAM               802
SET 12 KIDS COLOUR  CHALK STICKS      795
PACK OF 12 TRADITIONAL CRAYONS        795
HANGING METAL HEART LANTERN           795
ASSORTED COLOUR MINI CASES            770
TEA TIME OVEN GLOVE                   755
TRADITIONAL MODELLING CLAY            754
POPPY'S PLAYHOUSE KITCHEN             752
PAINTED METAL PEARS ASSORTED          747
ENAMEL BREAD BIN CREAM                740
SET 12 RETRO WHITE CHALK STICKS       725
HEART FILIGREE DOVE LARGE             724
PICTURE DOMINOES                      722
PINK HEART SHAPE EGG FRYING PAN       722
WHITE METAL LANTERN   

In [45]:
other_revenue_pct = df_sales[df_sales['Category'] == 'Other']['Revenue'].sum() / df_sales['Revenue'].sum() * 100
print(f"'Other' represents {other_revenue_pct:.2f}% of total revenue")

'Other' represents 26.22% of total revenue


In [46]:
def categorize_product(description):
    desc = str(description).upper()
    
    if any(word in desc for word in ['CHRISTMAS', 'XMAS', 'ADVENT']):
        return 'Christmas & Seasonal'
    elif any(word in desc for word in ['BAG', 'SHOPPER', 'BASKET']):
        return 'Bags & Storage'
    elif any(word in desc for word in ['CAKE', 'BAKING', 'TEATIME', 'MUG', 'TEACUP', 'SAUCER',
                                         'SNACK BOX', 'HOT WATER BOTTLE', 'KITCHEN SCALES',
                                         'JAM MAKING', 'COOKIE CUTTER', 'NAPKIN', 'LID GLASS BOWL',
                                         'JELLY MOULD', 'TEA SET', 'TEA GLASS', 'MEASURING SPOONS',
                                         'MILK JUG', 'DOILIES', 'ENAMEL', 'OVEN GLOVE', 'BREAD BIN',
                                         'FRYING PAN', 'BREAKFAST PLATE', 'PAPER PLATE', 'PAPER CUP',
                                         'FOOD CONTAINER']):
        return 'Kitchen & Dining'
    elif any(word in desc for word in ['LIGHT HOLDER', 'T-LIGHT', 'CANDLE', 'ORNAMENT', 'FRAME',
                                         'CHALKBOARD', 'NIGHT LIGHT', 'DECORATION', 'CHILLI LIGHTS',
                                         'POPCORN HOLDER', 'LANTERN', 'DOORSTOP', 'CUSHION COVER',
                                         'HEART']):
        return 'Home Decor & Lighting'
    elif 'SIGN' in desc:
        return 'Novelty Signs'
    elif any(word in desc for word in ['BUNTING', 'PARTY', 'BALLOON']):
        return 'Party & Occasion'
    elif any(word in desc for word in ['TRINKET', 'RECIPE BOX', 'TIN', 'BOX', 'CABINET', 'COAT RACK',
                                         'MINI CHEST']):
        return 'Storage & Boxes'
    elif any(word in desc for word in ['CARD', 'WRAP', 'GIFT TAG', 'RIBBON', 'PAPER CHAIN',
                                         'NOTEBOOK', 'TISSUE', 'PENCIL', 'STATIONERY']):
        return 'Stationery & Gift Wrap'
    elif any(word in desc for word in ['FELT', 'CRAFT', 'WICKER']):
        return 'Craft & Textile'
    elif any(word in desc for word in ['BUILDING BLOCK', 'ALARM CLOCK', 'HAND WARMER', 'PLASTERS',
                                         'PARASOL', 'PEG', 'KEY FOB', 'SEWING KIT', 'FAN', 'APRON',
                                         'HAIR CLIP', 'GLITTER']):
        return 'Novelty & Personal Items'
    elif any(word in desc for word in ['DOORMAT', 'DOOR MAT', 'GARDEN', 'PLANT LADDER', 'KNEELING PAD']):
        return 'Garden & Outdoor'
    elif any(word in desc for word in ['HOOK', 'HANGER', 'DRAWER KNOB']):
        return 'Hardware & Fixtures'
    elif any(word in desc for word in ['SPINNING TOP', 'SKIPPING ROPE', 'SNAKES & LADDERS',
                                         'BINGO', 'CATCH CUP', 'TOY', 'GAME', 'COLOURING',
                                         'CRAYON', 'CHALK STICK', 'MODELLING CLAY', 'PAINT SET',
                                         'DOMINOES', 'SKITTLES', 'PLAYHOUSE']):
        return 'Toys & Kids Craft'
    else:
        return 'Other'

df_sales['Category'] = df_sales['Description'].apply(categorize_product)
df_sales['Category'].value_counts(normalize=True) * 100

Category
Other                       21.035884
Home Decor & Lighting       16.495856
Kitchen & Dining            15.323503
Bags & Storage               9.927530
Storage & Boxes              7.345444
Stationery & Gift Wrap       6.676154
Novelty Signs                5.851700
Christmas & Seasonal         4.354169
Novelty & Personal Items     3.186998
Garden & Outdoor             2.781797
Party & Occasion             2.025608
Toys & Kids Craft            1.773179
Craft & Textile              1.720959
Hardware & Fixtures          1.501218
Name: proportion, dtype: float64

In [47]:
df_sales['Category'] = df_sales['Description'].apply(categorize_product)

print(df_sales['Category'].value_counts(normalize=True) * 100)

other_revenue_pct = df_sales[df_sales['Category'] == 'Other']['Revenue'].sum() / df_sales['Revenue'].sum() * 100
print(f"'Other' represents {other_revenue_pct:.2f}% of total revenue")

Category
Other                       21.035884
Home Decor & Lighting       16.495856
Kitchen & Dining            15.323503
Bags & Storage               9.927530
Storage & Boxes              7.345444
Stationery & Gift Wrap       6.676154
Novelty Signs                5.851700
Christmas & Seasonal         4.354169
Novelty & Personal Items     3.186998
Garden & Outdoor             2.781797
Party & Occasion             2.025608
Toys & Kids Craft            1.773179
Craft & Textile              1.720959
Hardware & Fixtures          1.501218
Name: proportion, dtype: float64
'Other' represents 19.16% of total revenue


In [48]:
df_sales[df_sales['Category'] == 'Other'].groupby('Description')['Revenue'].sum().sort_values(ascending=False).head(30)

Description
MEDIUM CERAMIC TOP STORAGE JAR         81700.92
WOOD BLACK BOARD ANT WHITE FINISH      56745.73
VINTAGE UNION JACK MEMOBOARD           45402.18
SET/4 WHITE RETRO STORAGE CUBES        43844.96
REGENCY TEAPOT ROSES                   24879.13
ASSORTED COLOUR MINI CASES             22097.61
CLASSIC METAL BIRDCAGE PLANT HOLDER    19194.13
PAINTED METAL PEARS ASSORTED           17451.45
RED RETROSPOT TRADITIONAL TEAPOT       17030.56
CHILDRENS CUTLERY POLKADOT PINK        16865.97
SET OF 3 COLOURED  FLYING DUCKS        16218.10
CHILDS BREAKFAST SET SPACEBOY          15946.55
CHILDS BREAKFAST SET DOLLY GIRL        15795.01
RED RETROSPOT UMBRELLA                 15791.12
PINK 3 PIECE MINI DOTS CUTLERY SET     15767.15
RED 3 PIECE MINI DOTS CUTLERY SET      15738.53
RED DINER WALL CLOCK                   15658.98
RED/WHITE DOTS RUFFLED UMBRELLA        15476.88
RED RETROSPOT MINI CASES               15118.47
ZINC FOLKART SLEIGH BELLS              15072.58
BROCADE RING PURSE          

In [49]:
def categorize_product(description):
    desc = str(description).upper()
    
    if any(word in desc for word in ['CHRISTMAS', 'XMAS', 'ADVENT']):
        return 'Christmas & Seasonal'
    elif any(word in desc for word in ['BAG', 'SHOPPER', 'BASKET']):
        return 'Bags & Storage'
    elif any(word in desc for word in ['CAKE', 'BAKING', 'TEATIME', 'MUG', 'TEACUP', 'SAUCER',
                                         'SNACK BOX', 'HOT WATER BOTTLE', 'KITCHEN SCALES',
                                         'JAM MAKING', 'COOKIE CUTTER', 'NAPKIN', 'LID GLASS BOWL',
                                         'JELLY MOULD', 'TEA SET', 'TEA GLASS', 'MEASURING SPOONS',
                                         'MILK JUG', 'DOILIES', 'ENAMEL', 'OVEN GLOVE', 'BREAD BIN',
                                         'FRYING PAN', 'BREAKFAST PLATE', 'PAPER PLATE', 'PAPER CUP',
                                         'FOOD CONTAINER', 'TEAPOT', 'COFFEE SET', 'TEA TOWEL',
                                         'CUTLERY', 'BREAKFAST SET']):
        return 'Kitchen & Dining'
    elif any(word in desc for word in ['LIGHT HOLDER', 'T-LIGHT', 'CANDLE', 'ORNAMENT', 'FRAME',
                                         'CHALKBOARD', 'NIGHT LIGHT', 'DECORATION', 'CHILLI LIGHTS',
                                         'POPCORN HOLDER', 'LANTERN', 'DOORSTOP', 'CUSHION COVER',
                                         'HEART', 'MEMOBOARD', 'BIRDCAGE', 'PLANT HOLDER',
                                         'CHERRY LIGHTS', 'WALL CLOCK']):
        return 'Home Decor & Lighting'
    elif 'SIGN' in desc:
        return 'Novelty Signs'
    elif any(word in desc for word in ['BUNTING', 'PARTY', 'BALLOON']):
        return 'Party & Occasion'
    elif any(word in desc for word in ['TRINKET', 'RECIPE BOX', 'TIN', 'BOX', 'CABINET', 'COAT RACK',
                                         'MINI CHEST', 'STORAGE JAR', 'STORAGE CUBE', 'CRATE',
                                         'MINI CASES']):
        return 'Storage & Boxes'
    elif any(word in desc for word in ['CARD', 'WRAP', 'GIFT TAG', 'RIBBON', 'PAPER CHAIN',
                                         'NOTEBOOK', 'TISSUE', 'PENCIL', 'STATIONERY']):
        return 'Stationery & Gift Wrap'
    elif any(word in desc for word in ['FELT', 'CRAFT', 'WICKER']):
        return 'Craft & Textile'
    elif any(word in desc for word in ['BUILDING BLOCK', 'ALARM CLOCK', 'HAND WARMER', 'PLASTERS',
                                         'PARASOL', 'PEG', 'KEY FOB', 'SEWING KIT', 'FAN', 'APRON',
                                         'HAIR CLIP', 'GLITTER', 'UMBRELLA', 'PURSE', 'LIP GLOSS']):
        return 'Novelty & Personal Items'
    elif any(word in desc for word in ['DOORMAT', 'DOOR MAT', 'GARDEN', 'PLANT LADDER', 'KNEELING PAD',
                                         'GROW YOUR OWN']):
        return 'Garden & Outdoor'
    elif any(word in desc for word in ['HOOK', 'HANGER', 'DRAWER KNOB']):
        return 'Hardware & Fixtures'
    elif any(word in desc for word in ['SPINNING TOP', 'SKIPPING ROPE', 'SNAKES & LADDERS',
                                         'BINGO', 'CATCH CUP', 'TOY', 'GAME', 'COLOURING',
                                         'CRAYON', 'CHALK STICK', 'MODELLING CLAY', 'PAINT SET',
                                         'DOMINOES', 'SKITTLES', 'PLAYHOUSE', 'FLYING DUCKS']):
        return 'Toys & Kids Craft'
    else:
        return 'Other'

df_sales['Category'] = df_sales['Description'].apply(categorize_product)

print(df_sales['Category'].value_counts(normalize=True) * 100)
other_revenue_pct = df_sales[df_sales['Category'] == 'Other']['Revenue'].sum() / df_sales['Revenue'].sum() * 100
print(f"'Other' represents {other_revenue_pct:.2f}% of total revenue")

Category
Other                       17.989301
Kitchen & Dining            17.069974
Home Decor & Lighting       16.982576
Bags & Storage               9.927530
Storage & Boxes              7.358100
Stationery & Gift Wrap       6.676154
Novelty Signs                5.736398
Christmas & Seasonal         4.354169
Novelty & Personal Items     3.942491
Garden & Outdoor             2.862419
Party & Occasion             2.025608
Toys & Kids Craft            1.853103
Craft & Textile              1.720959
Hardware & Fixtures          1.501218
Name: proportion, dtype: float64
'Other' represents 14.02% of total revenue


In [50]:
df_sales[df_sales['Category'] == 'Other'].groupby('Description')['Revenue'].sum().sort_values(ascending=False).head(50)

Description
WOOD BLACK BOARD ANT WHITE FINISH     56745.73
PAINTED METAL PEARS ASSORTED          17451.45
ZINC FOLKART SLEIGH BELLS             15072.58
WOODEN HAPPY BIRTHDAY GARLAND         14593.32
SET OF 10 LED DOLLY LIGHTS            13290.52
LARGE CIRCULAR MIRROR MOBILE          13153.80
SET 3 RETROSPOT TEA,COFFEE,SUGAR      12764.91
SOMBRERO                              12450.85
REGENCY SUGAR BOWL GREEN              12342.43
ANTIQUE LILY FAIRY LIGHTS             12234.56
ASSORTED FLOWER COLOUR "LEIS"         11961.49
BABUSHKA LIGHTS STRING OF 10          11542.60
PIGGY BANK RETROSPOT                  11492.82
SOLDIERS EGG CUP                      11238.45
UNION STRIPE WITH FRINGE  HAMMOCK     11199.18
REGENCY TEA PLATE ROSES               11097.25
CAMOUFLAGE LED TORCH                  10681.67
SET 10 LIGHTS NIGHT OWL               10509.88
10 COLOUR SPACEBOY PEN                10397.33
PHOTO CUBE                            10374.33
REVOLVER WOODEN RULER                 10373.99
W

In [51]:
df_sales[df_sales['Description'] == 'Adjust bad debt']['StockCode'].unique()

array([], dtype=object)

In [52]:
df_sales[df_sales['Category'] == 'Other'].groupby('Description')['Revenue'].sum().sort_values(ascending=False).head(50)

Description
WOOD BLACK BOARD ANT WHITE FINISH     56745.73
PAINTED METAL PEARS ASSORTED          17451.45
ZINC FOLKART SLEIGH BELLS             15072.58
WOODEN HAPPY BIRTHDAY GARLAND         14593.32
SET OF 10 LED DOLLY LIGHTS            13290.52
LARGE CIRCULAR MIRROR MOBILE          13153.80
SET 3 RETROSPOT TEA,COFFEE,SUGAR      12764.91
SOMBRERO                              12450.85
REGENCY SUGAR BOWL GREEN              12342.43
ANTIQUE LILY FAIRY LIGHTS             12234.56
ASSORTED FLOWER COLOUR "LEIS"         11961.49
BABUSHKA LIGHTS STRING OF 10          11542.60
PIGGY BANK RETROSPOT                  11492.82
SOLDIERS EGG CUP                      11238.45
UNION STRIPE WITH FRINGE  HAMMOCK     11199.18
REGENCY TEA PLATE ROSES               11097.25
CAMOUFLAGE LED TORCH                  10681.67
SET 10 LIGHTS NIGHT OWL               10509.88
10 COLOUR SPACEBOY PEN                10397.33
PHOTO CUBE                            10374.33
REVOLVER WOODEN RULER                 10373.99
W

In [53]:
def categorize_product(description):
    desc = str(description).upper()
    
    if any(word in desc for word in ['CHRISTMAS', 'XMAS', 'ADVENT']):
        return 'Christmas & Seasonal'
    elif any(word in desc for word in ['BAG', 'SHOPPER', 'BASKET']):
        return 'Bags & Storage'
    elif any(word in desc for word in ['CAKE', 'BAKING', 'TEATIME', 'MUG', 'TEACUP', 'SAUCER',
                                         'SNACK BOX', 'HOT WATER BOTTLE', 'KITCHEN SCALES',
                                         'JAM MAKING', 'COOKIE CUTTER', 'NAPKIN', 'LID GLASS BOWL',
                                         'JELLY MOULD', 'TEA SET', 'TEA GLASS', 'MEASURING SPOONS',
                                         'MILK JUG', 'DOILIES', 'ENAMEL', 'OVEN GLOVE', 'BREAD BIN',
                                         'FRYING PAN', 'BREAKFAST PLATE', 'PAPER PLATE', 'PAPER CUP',
                                         'FOOD CONTAINER', 'TEAPOT', 'COFFEE SET', 'TEA TOWEL',
                                         'CUTLERY', 'BREAKFAST SET', 'SUGAR BOWL', 'TEA PLATE',
                                         'BUTTER DISH', 'BEURRE DISH', 'EGG CUP', 'EGG HOLDER',
                                         'ROLLING PIN', 'TEA CADDY', 'TABLE CLOTH', 'SUGAR JAM BOWL',
                                         'BOWL', 'GLASS JAR', 'CONTAINER SET', 'WASHING UP',
                                         'CHOPPING BOARD', 'PLACEMAT', 'PLATE', 'COOKING SET',
                                         'TEA,COFFEE,SUGAR', 'TUMBLERS', 'MILK PAN', 'SCISSOR']):
        return 'Kitchen & Dining'
    elif any(word in desc for word in ['LIGHT HOLDER', 'T-LIGHT', 'CANDLE', 'ORNAMENT', 'FRAME',
                                         'CHALKBOARD', 'BLACK BOARD', 'NIGHT LIGHT', 'NIGHTLIGHT', 'DECORATION',
                                         'POPCORN HOLDER', 'LANTERN', 'DOORSTOP',
                                         'CUSHION', 'HEART', 'MEMOBOARD', 'BIRDCAGE',
                                         'PLANT HOLDER', 'WALL CLOCK',
                                         'MIRROR', 'DISCO BALL',
                                         'TORCH', 'MOBILE', 'WALL ART', 'PHOTO CUBE', 'SLATE TILE',
                                         'LIGHTS', 'PAINTED METAL', 'SLEIGH BELLS', 'LAMP',
                                         'PAINTED WOOD', 'HONEYPOT', 'DRAWING SLATE', 'TV DINNER TRAY']):
        return 'Home Decor & Lighting'
    elif 'SIGN' in desc:
        return 'Novelty Signs'
    elif any(word in desc for word in ['BUNTING', 'PARTY', 'BALLOON', 'GARLAND', 'SOMBRERO', 'LEIS']):
        return 'Party & Occasion'
    elif any(word in desc for word in ['TRINKET', 'RECIPE BOX', 'TIN', 'BOX', 'CABINET', 'COAT RACK',
                                         'MINI CHEST', 'STORAGE JAR', 'STORAGE CUBE', 'CRATE',
                                         'MINI CASES', 'ROUND CONTAINER', 'MAGAZINE RACK',
                                         'SIDEBOARD', 'ORGANISER', 'DRAWER', 'WALL TIDY',
                                         'WALL MOUNTED', 'JEWELLERY DRAWER']):
        return 'Storage & Boxes'
    elif any(word in desc for word in ['CARD', 'WRAP', 'GIFT TAG', 'RIBBON', 'PAPER CHAIN',
                                         'NOTEBOOK', 'TISSUE', 'PENCIL', 'STATIONERY', 'PEN',
                                         'RULER', 'EXERCISE BOOK']):
        return 'Stationery & Gift Wrap'
    elif any(word in desc for word in ['FELT', 'CRAFT', 'WICKER', 'PATCHES', 'PAINT YOUR OWN']):
        return 'Craft & Textile'
    elif any(word in desc for word in ['BUILDING BLOCK', 'ALARM CLOCK', 'HAND WARMER', 'PLASTERS',
                                         'PARASOL', 'PEG', 'KEY FOB', 'SEWING KIT', 'FAN', 'APRON',
                                         'HAIR CLIP', 'GLITTER', 'UMBRELLA', 'PURSE', 'LIP GLOSS',
                                         'PIGGY BANK', 'BABY GIFT', 'PONCHO', 'SLIPPER', 'HAMMOCK',
                                         'BLOCK LETTERS', 'WOOD LETTERS', 'FRIDGE MAGNET', 'MAGNET',
                                         'CALCULATOR', 'MATCHES', 'FLANNEL']):
        return 'Novelty & Personal Items'
    elif any(word in desc for word in ['DOORMAT', 'DOOR MAT', 'GARDEN', 'PLANT LADDER', 'KNEELING PAD',
                                         'GROW YOUR OWN', 'WATERING CAN', 'WHEELBARROW', 'WINDMILL',
                                         'ACAPULCO MAT']):
        return 'Garden & Outdoor'
    elif any(word in desc for word in ['HOOK', 'HANGER', 'DRAWER KNOB']):
        return 'Hardware & Fixtures'
    elif any(word in desc for word in ['SPINNING TOP', 'SKIPPING ROPE', 'SNAKES & LADDERS',
                                         'BINGO', 'CATCH CUP', 'TOY', 'GAME', 'COLOURING',
                                         'CRAYON', 'CHALK STICK', 'MODELLING CLAY', 'PAINT SET',
                                         'DOMINOES', 'SKITTLES', 'PLAYHOUSE', 'FLYING DUCKS', 'LUDO',
                                         'NAUGHTS & CROSSES', 'INFLATABLE', 'GLOBE', 'TEDDY BEAR',
                                         'BEACH SPADE']):
        return 'Toys & Kids Craft'
    else:
        return 'Other'

df_sales['Category'] = df_sales['Description'].apply(categorize_product)

print(df_sales['Category'].value_counts(normalize=True) * 100)
other_revenue_pct = df_sales[df_sales['Category'] == 'Other']['Revenue'].sum() / df_sales['Revenue'].sum() * 100
print(f"'Other' represents {other_revenue_pct:.2f}% of total revenue")

Category
Kitchen & Dining            20.185819
Home Decor & Lighting       19.128767
Bags & Storage               9.927530
Other                        9.549435
Storage & Boxes              8.266763
Stationery & Gift Wrap       7.471309
Novelty Signs                5.513368
Novelty & Personal Items     5.244994
Christmas & Seasonal         4.354169
Garden & Outdoor             3.292933
Party & Occasion             2.704962
Toys & Kids Craft            2.052116
Craft & Textile              1.564599
Hardware & Fixtures          0.743234
Name: proportion, dtype: float64
'Other' represents 5.86% of total revenue


In [54]:
df_sales.to_csv('../data/processed/online_retail_cleaned.csv', index=False)